### PR TITLE
When a project or document is removed/closed, clear out all the squiggles we currently have for them.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This affects TS as they routlinely close/open projects behind the scenes, while still using the
current editor buffer.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
retest prtest/win/dbg/eta  please

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs [hunk 1]
```diff
@@ -198,6 +210,52 @@ private void OnDiagnosticsUpdated(DiagnosticsUpdatedArgs e)
                 }
             }
 
+            private void ProcessDocumentOrProjectRemoval(DiagnosticsUpdatedArgs e)
+            {
+                if (e.Solution != null)
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: That's a bit subtle: e.Solution == null is the only way to know it's a project removal?

Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: See also [this](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs,144), which implies you might also want to look at Document?

**Comment by Pilchie** [comment]
Also tagging @heejaechang and @mavasani

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 1 above)
Comment (heejaechang) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: up to suggestion. the args could have a property saying whether this is removal or not. or we could set the solution and then ask caller to check solution.GetDocument(documentId) == null.

**Comment by heejaechang** [comment]
:+1:

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 1 above)
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: +1 on implementing that suggestion, a property making it explicit would be nice.

**Comment by mavasani** [comment]
:+1:

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs]: +1 on that suggestion as well.  I'll see about making that as part of this change.

**Comment by CyrusNajmabadi** [comment]
Ok.  I've added a 'Kind' to let you know if diagnostics are being removed, or if they're being created.

Please let me know if this looks ok to you @heejaechang and @mavasani

**Comment by heejaechang** [comment]
:+1:

File: src/Features/Core/Portable/Diagnostics/DiagnosticsUpdatedArgs.cs [hunk 2]
```diff
@@ -1,21 +1,76 @@
...
         }
+
+        public static DiagnosticsUpdatedArgs DiagnosticsRemoved(
+            object id,
+            Workspace workspace,
+            Solution solution,
+            ProjectId projectId,
+            DocumentId documentId)
+        {
+            return new DiagnosticsUpdatedArgs(id, workspace, solution, projectId, documentId, ImmutableArray<DiagnosticData>.Empty, DiagnosticsUpdatedKind.DiagnosticsCreated);
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/DiagnosticsUpdatedArgs.cs]: Typo: Shouldn't the argument here be DiagnosticsUpdatedKind.DiagnosticsRemoved?

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/DiagnosticsUpdatedArgs.cs]: Yup yup!

**Comment by Pilchie** [comment]
:+1:

**Comment by CyrusNajmabadi** [comment]
Ok.  I'm working with @vladima on validating this change.  One validated. I'll get sign off to push into Stabilization.

**Comment by CyrusNajmabadi** [comment]
retest prtest/win/dbg/unit32 please

**Comment by CyrusNajmabadi** [comment]
retest prtest/win/dbg/unit64 please

**Comment by vladima** [comment]
we've validated that on basic scenarios (including the repro case) we now have correct behavior

**Comment by CyrusNajmabadi** [comment]
Note: without this fix, squiggles in TypeScript will stay around until VS restart during very normal/common TS editing operations.  For example,

if the user changes a "///<reference...>" line (equivalent to our "import namespace"), they can end up with squiggles that will never go away until VS restart.

**Pull request merged**